src/index.tsx

```diff
@@ -62,24 +98,29 @@ export const createPortalNode = <C extends Component<any>>(): PortalNode<C> => {
             if (parent && lastPlaceholder) {
                 parent.replaceChild(
                     lastPlaceholder,
-                    portalNode
+                    portalNode.element,
                 );
 
                 parent = undefined;
                 lastPlaceholder = undefined;
             }
         }
-    });
+    } as AnyPortalNode<C>;
 
     return portalNode;
 };
 
-export class InPortal extends React.PureComponent<InPortalProps, { nodeProps: {} }> {
+interface InPortalProps {
+    node: AnyPortalNode;
+    children: React.ReactNode;
+}
+
+class InPortal extends React.PureComponent<InPortalProps, { nodeProps: {} }> {
```

Comment:
When testing in a sandbox, this was quick to build and worked exactly as you describe: `InPortal` can check each of its children after mount or update to ensure they're the right type, and throw an error if there's a mismatch.

Unfortunately, with the `<circle>` example here the browser doesn't create an SVG element at all: it emits a "The tag <circle> is unrecognized in this browser." warning and creates a HTML element instead -- a normal, valid, well-formed HTML element -- which then passes the check.

I did some digging into React to see exactly how they detect SVG, and under the hood it's done by persisting whatever `parentNamespace` the element was rendered within: [ReactDOMComponent.js#L414-L418](https://github.com/facebook/react/blob/62861bbcc752c63728592fdc45078854f9aae161/packages/react-dom/src/client/ReactDOMComponent.js#L414-L418). The value of the `parentNamespace` argument ultimately comes from the Fiber's reference to its parent, at [ReactFiberHostContext.js#L73](https://github.com/facebook/react/blob/42b75ab007a5e7c159933cfdbf2b6845d89fc7f2/packages/react-reconciler/src/ReactFiberHostContext.js#L73) -- but in the `<circle>` example even that is still HTML, not SVG, so even React doesn't know that it's wrong. I verified this with some plain output, without portals, and it's the same.

So, I don't think it's feasible to detect elements which are _supposed to be_ SVG but aren't.

InPortal's childNodes check does work when rendering a namespace-changing element, though -- so we can detect when a `<svg>` is rendered directly into a HTML portal.

**Except that that works fine!**
```
const portalNode = createHtmlPortalNode();

// This works!
<InPortal node={portalNode}>
  <svg>
    <circle cx="50" cy="50" r="50" />
  </svg>
</InPortal>

<div>
  Portal renders here:
  <OutPortal node={portalNode} />
</div>
```

React has a bunch of tests around the element namespaces in portals: [ReactDOMFiber-test.js#L500](https://github.com/facebook/react/blob/60016c448bb7d19fc989acd05dda5aca2e124381/packages/react-dom/src/__tests__/ReactDOMFiber-test.js#L500) -- and they all seem to just ensure that the parentNamespace tracking works the same through portals as it does normally.

========

So, after all that:
* The root cause behind ALL of this svg-support stuff is the intermediate element we create for the portal node: it has to be created in the proper namespace for React, not for the browser.
    * If we render a plain `document.createElement('div')` for the portal node then React resets the namespace from SVG to HTML, and then uses that for its nested elements -- and that's the reason SVG elements weren't working initially.
* The earlier `InPortal`/`OutPortal` order issues were due to React checking the `parentNamespace` at `ReactDOM.createElement()` time -- that's why the element had to have the correct namespace at that point.

I have not found any way to avoid the intermediate portal node element. (More on that below, for textNodes, but based on my testing it seems like it's unavoidable.) If we could find a way to avoid it, though, and only render the _direct_ `childNodes` sent into `InPortal` (without needing to add a `<div>` or `<g>` around them for the portal node element) then we wouldn't need to have different portal types: React would keep the namespaces intact through the portal just as it does normally.

Reply:
I reverted the `InPortal` check in commit [a5b9379](https://github.com/spautz/react-reverse-portal/commit/a5b9379909666b5faac4929e0b796b47844ffb72) based on this, since it didn't stop the accidental error case, but does stop the case that works.

Reply:
Very thorough, makes sense, I'm convinced. Sounds like browsers will warn about this case anyway and it's not essential, let's skip this.